popde/adaptive_kde.py

```diff
@@ -0,0 +1,191 @@
+    sample = np.column_stack((sample1, sample2, sample3)) # shape is (n_points, n_features)
+    # Create and fit the adaptive KDE note backend must be set correctly
+    kde = AdaptiveBwKDE(sample, backend='awKDEpy', dim_names=['x', 'y', 'z'], alpha=0.5, input_transf=None)
+    print("kde=", kde)
+    
+    # Generate grid for plotting
+    xgrid = np.linspace(sample1.min(), sample1.max(), 100)
+    ygrid = np.linspace(sample2.min(), sample2.max(), 100)
+    zgrid = np.linspace(sample3.min(), sample3.max(), 100)
+    XX, YY, ZZ = np.meshgrid(xgrid, ygrid, zgrid)
+    eval_pts = np.column_stack((XX.flatten(), YY.flatten(), ZZ.flatten()))
+    kde.fit()
+
+    # Evaluate the KDE at the grid points
+    density_values = kde.evaluate(eval_pts)
+    density_values = density_values.reshape(XX.shape)
+    print(density_values)
+    """
+    def __init__(self, data, backend='KDEpy', bandwidth=1., dim_names=None, alpha=0.5,  input_transf=None, stdize=False, rescale=None):
+        # Additional parameter for the Wang & Wang formula
+        self.alpha = alpha
+        super().__init__(data, input_transf, stdize, rescale,
+                         backend, bandwidth, dim_names)  
+
+    def _calc_local_bandwidth(self, kde_values):
+        """
+        Calculate local bandwidth using the provided function.
+
+        Parameters:
+        -----------
+        kde_values : array-like
+            The KDE values at the data point positions.
+
+        Returns:
+        --------
+        inv_loc_bw : array-like
+            Inverse of the local bandwidth calculated using the provided function.
+        """
+        g = np.exp(np.sum(np.log(kde_values)) / len(kde_values))
+        inv_loc_bw = (kde_values / g) ** self.alpha
+        return inv_loc_bw
+
+    def calculate_per_point_bandwidths(self, pilot_values, adaptive=True):
+        """
+        Calculate per-point bandwidths using the Wang & Wang formula.
+
+        Parameters:
+        -----------
+        pilot_values : array-like
+            The pilot KDE values at the data point positions.
+        adaptive : bool, optional (default=True)
+            If True, use adaptive bandwidth; if False, use fixed bandwidth.
+
+        Returns:
+        --------
+        per_point_bandwidths : array-like
+            Per-point bandwidths calculated using the Wang & Wang formula.
+        """
+        # Calculate the local bandwidth using the provided function
+        local_bandwidths = self._calc_local_bandwidth(pilot_values)
+
+        # Use the local bandwidths to calculate per-point bandwidths
+        if adaptive:
+            per_point_bandwidths = self.bandwidth * local_bandwidths
+        else:
+            per_point_bandwidths = np.ones_like(local_bandwidths) / self.bandwidth
+
+        return per_point_bandwidths
+
+    def fit_awKDEpy(self):
+        """
+        Fit the adaptive KDE
+        """
+        pilot_kde = TreeKDE(bw=self.bandwidth).fit(self.kde_data)
```

Comment:
do you still need to fit a pilot KDE here? Doesn't `VariableBwKDEPy.__init__` already use the fit() method, and so the output of this is already present and could be used as the pilot KDE?

Reply:
That is not clear to me how pilot kde can be obtained from superclass here

Reply:
well, look at the VariableBwKDEPy.__init__ method.  It does `self.fit()`.  That calls `fit_method` which here is `fit_KDEpy`, which sets `self.kernel_estimate` to be a TreeKDE object (line 193 of density_estimate.py).  So then you can use `self.kernel_estimate`, no?

In general, you should try to use methods or attributes that already exist to avoid duplication.

Reply:
I am assuming you read the code for the superclass, if you see a specific issue there can you mention it, otherwise I can't tell what may not be clear.

Reply:
Using 
 `pilot_kde = self.fit()`
I got an error:

> density_estimate.py", line 126, in fit
>     fit_method()
> RecursionError: maximum recursion depth exceeded

Reply:
work in progress

Reply:
ah, so what problem might this indicate?

Reply:
so I did not say you had to call self.fit() again, because that is already called in the __ini__ method of the parent class.  I said you could use self.kernel_estimate ..

Reply:
in general, you can put in print statements to see what is going wrong, eg you could put in `print(fit_method)` or `print(self.backend)`

Reply:
no, I did not push the failed script on PR. Also self.estimate give error,  
work in progress

Reply:
issue maybe in setting  backened option.

Reply:
this commit did not work but commit the change.
fix required